runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -606,8 +606,13 @@ static void *dereferenceStructPointerChain(void *baseStruct, TR::Node *baseNode,
                   {
                   // if stable array, check that we are loading from the beginning of an element that it's not Null
                   // size of the element was checked in verifyFieldAccess
-                  if (fieldAddress % symRef->getSymbol()->getSize() != 0 ||
-                      isNullValueAtAddress(comp, symRef->getSymbol()->getDataType(), fieldAddress, symRef->getSymbol()))
+                  TR::DataType type = symRef->getSymbol()->getDataType();
+                  int32_t width = TR::Symbol::convertTypeToSize(type);
+                  if (comp->useCompressedPointers() && type == TR::Address)
+                     width = TR::Compiler->om.sizeofReferenceField();
+
+                  if ((fieldAddress % width) != 0 ||
+                      isNullValueAtAddress(comp, type, fieldAddress, symRef->getSymbol()))
```

Comment:
Does this mean that the checking inside `isNullValueAtAddress` also is aware of compressed refs ? i.e. it will only check that 4 bytes are null ?

Reply:
Interestingly, it has this code:

```
  else if (comp->getSymRefTab()->isImmutableArrayShadow(field) || isStableArrayElement)
      {
      TR_OpaqueClassBlock *arrayClass = fej9->getObjectClass((uintptr_t)curStruct);
      if (!fej9->isClassArray(arrayClass) ||
          (field->getSymbol()->isCollectedReference() &&
          fej9->isPrimitiveArray(arrayClass)) ||
          (!field->getSymbol()->isCollectedReference() &&
          fej9->isReferenceArray(arrayClass)))
         return false;

      if (isStableArrayElement && fej9->isPrimitiveArray(arrayClass))
         {
         if (TR::Compiler->om.getArrayElementWidthInBytes(comp, (uintptr_t)curStruct) != field->getSymbol()->getSize())
            return false;
         }
      return true;
      }
```

And I am not sure why it only checks primitive arrays...

Reply:
Yes, [`isNullValueAtAddress()` uses `getReferenceFieldAtAddress()`](https://github.com/eclipse-openj9/openj9/blob/e102f308bbde5dfe560fc37eb9b1732565aff5a2/runtime/compiler/optimizer/J9TransformUtil.cpp#L240-L247)

Reply:
@gita-omr, the code you pasted is from `verifyFieldAccess()`, not `isNullValueAtAddress()`, so I don't think it's directly relevant to Vijay's question.

Looking back at #13926, it seems that `getArrayElementWidthInBytes(...) == getSymbol()->getSize()` is required there only for primitive arrays because it's just there to [guard against the possibility of loading a primitive type that is larger than the primitive array elements](https://github.com/eclipse-openj9/openj9/pull/13926#discussion_r757701997) and which can therefore be nonzero even if it overlaps a zero-valued array element which is still mutable. We could have also checked this for references, but we only load references from reference arrays anyway, so the alignment test (here in `dereferenceStructPointerChain()`) is sufficient. Under compressed refs, comparing to `getSymbol()->getSize()` would just [fail and spuriously prevent folding](https://github.com/eclipse-openj9/openj9/pull/13926#discussion_r757744157):
> In the reference case, the sizes will appear to mismatch with compressed references, but it will actually be ok

just like what happened here with elements at odd indices.

Really, we should always check that the offset is aligned when loading a reference from an array, not only when it's a stable array, and it's not obvious to me that we do, but if we don't, that problem is not especially related to this PR

Reply:
Thanks a lot for the clarification!